File: university/estrutura_de_dados/code/Ordenacao.java
public class Ordenacao {

    // -------------------------- counting sort --------------------------------
    // cost:
    static void countingSort(int[] arr) {
    }

    // -------------------------- quick sort -----------------------------------
    // cost:
    static void quickSort(int[] arr, int start_pointer, int end_pointer) {
        if (start_pointer < end_pointer) {
            int indexPivot = partition(arr, start_pointer, end_pointer);

            quickSort(arr, start_pointer, indexPivot - 1);
            quickSort(arr, indexPivot + 1, end_pointer);
        }
    }

    static int partition(int[] arr, int start_pointer, int end_pointer) {
        int pivot = arr[end_pointer];
        int i = start_pointer;

        for (int j = start_pointer; j < end_pointer; j++) {
            if (arr[j] <= pivot) {
                int aux = arr[j];
                arr[j] = arr[i];
                arr[i] = aux;
                i++;
            }
        }

        int aux = arr[i];
        arr[i] = arr[end_pointer];
        arr[end_pointer] = aux;

        return i;
    }

    // -------------------------- merge sort -------------------------------
    // cost:
    static void mergeSort(int[] arr, int end_pointer) {
        if (end_pointer < 2) {
            return;
        }
        int mid = end_pointer / 2;
        int[] l = new int[mid];
        int[] r = new int[end_pointer - mid];

        for (int i = 0; i < mid; i++) {
            l[i] = arr[i];
        }
        for (int i = mid; i < end_pointer; i++) {
            r[i - mid] = arr[i];
        }
        mergeSort(l, mid);
        mergeSort(r, end_pointer - mid);

        merge(arr, l, r, mid, end_pointer - mid);
    }

    static void merge(int[] arr, int[] l, int[] r, int left, int right) {
        int i = 0, j = 0, k = 0;
        while (i < left && j < right) {
            if (l[i] <= r[j]) {
                arr[k++] = l[i++];
            } else {
                arr[k++] = r[j++];
            }
        }
        while (i < left) {
            arr[k++] = l[i++];
        }
        while (j < right) {
            arr[k++] = r[j++];
        }
    }

    // -------------------------- insertion sort -------------------------------
    // cost:
    static void insertionSort(int[] arr) {
        int size = arr.length;

        for (int i = 1; i < size; i++) {
            int key = arr[i];
            int j = i - 1;

            while (j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j];
                j--;
            }

            arr[j + 1] = key;
        }
    }

    // -------------------------- bubble sort -------------------------------
    // cost:
    static void bubbleSort(int[] arr) {
        int size = arr.length;

        for (int j = 0; j < size - 1; j++) {
            for (int i = 0; i < size - 1; i++) {
                if (arr[i] > arr[i + 1]) {
                    int aux = arr[i];
                    arr[i] = arr[i + 1];
                    arr[i + 1] = aux;
                }
            }
        }
    }

    // -------------------------- selection sort -------------------------------
    // cost:
    static void selectionSort(int[] arr) {
        int size = arr.length;

        for (int j = 0; j < size; j++) {
            int indexMin = j;

            for (int i = j; i < size; i++) {
                if (arr[i] < arr[indexMin]) {
                    indexMin = i;
                }
            }

            if (arr[j] > arr[indexMin]) {
                int aux = arr[j];
                arr[j] = arr[indexMin];
                arr[indexMin] = aux;
            }
        }
    }

    // -------------------------- main -----------------------------------------
    public static void main(String[] args) {
        int[] array = { 3, 9, 2, 5, 1, 7 };

        mergeSort(array, array.length);

        System.out.println("\nOrdened array:");

        for (int x : array) {
            System.out.print(x + " ");
        }
    }
}

File: university/estrutura_de_dados/code/ordenacao.py
# ------------------------------ counting sort ------------------------------------
# cost: O(n)
def counting_sort(arr):
    size = len(arr)
    output = [0] * size

    count = [0] * 10

    for i in range(0, size):
        count[arr[i]] += 1

    for i in range(1, 10):
        count[i] += count[i - 1]

    i = size - 1
    while i >= 0:
        output[count[arr[i]] - 1] = arr[i]
        count[arr[i]] -= 1
        i -= 1

    for i in range(0, size):
        arr[i] = output[i]

# ------------------------------ quick sort ------------------------------------
# cost: O(n*log(n)) --> pivo aleatorio | O(n²) --> pivo ultimo elemento
def quick_sort(arr, start_pointer=0, end_pointer=None):
    if end_pointer is None:
        end_pointer = len(arr) - 1
    
    if start_pointer < end_pointer:
        index_pivot = partition(arr, start_pointer, end_pointer)
        quick_sort(arr, start_pointer, index_pivot-1)
        quick_sort(arr, index_pivot+1, end_pointer)


def partition(arr, start_pointer, end_pointer):
    pivot = arr[end_pointer]
    i = start_pointer
    
    for j in range(start_pointer, end_pointer):
        if arr[j] <= pivot:
            arr[j], arr[i] = arr[i], arr[j]
            i += 1
    
    arr[i], arr[end_pointer] = arr[end_pointer], arr[i]
    
    return i


# ------------------------------ merge sort ------------------------------------
# cost: O(n*log(n))
def merge_sort(arr, start_pointer=0, end_pointer=None):
    if end_pointer is None:
        end_pointer = len(arr)
   
    if (end_pointer - start_pointer > 1):
        mid_pointer = (end_pointer + start_pointer) // 2
        merge_sort(arr, start_pointer, mid_pointer)
        merge_sort(arr, mid_pointer, end_pointer)
        merge(arr, start_pointer, mid_pointer, end_pointer)


def merge(arr, start_pointer, mid_pointer, end_pointer):
    left = arr[start_pointer:mid_pointer]
    right = arr[mid_pointer:end_pointer]
    top_left, top_right = 0, 0
    
    for k in range(start_pointer, end_pointer):
        if top_left >= len(left):
            arr[k] = right[top_right]
            top_right += 1
        elif top_right >= len(right):
            arr[k] = left[top_left]
            top_left += 1
        elif left[top_left] < right[top_right]:
            arr[k] = left[top_left]
            top_left += 1
        else:
            arr[k] = right[top_right]
            top_right += 1


# ------------------------------ insertion sort --------------------------------
# cost: O(n²)
def insertion_sort(arr):
    size = len(arr)
    
    for i in range(1, size):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j+1] = arr[j]
            j -= 1
        arr[j+1] = key


# ------------------------------ bubble sort -----------------------------------
# cost: O(n²)
def bubble_sort(arr):
    size = len(arr)
    
    for j in range(size-1):
        for i in range(size-1):
            if arr[i] > arr[i+1]:
                arr[i], arr[i+1] = arr[i+1], arr[i]


# ------------------------------ selection sort --------------------------------
# cost: O(n²)
def selection_sort(arr):
    size = len(arr)
    
    for j in range(size):
        index_min = j
        for i in range(j, size):
            if arr[i] < arr[index_min]:
                index_min = i

        if arr[j] > arr[index_min]:
            arr[j], arr[index_min] = arr[index_min], arr[j]


# ---------------------------- main --------------------------------------------
def main():
    array = [3,9,2,5,1,7]

    quick_sort(array)
    
    print("\nOrdened array:")
    print(array)


if __name__=='__main__':
    main()